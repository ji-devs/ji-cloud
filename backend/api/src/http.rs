use std::{net::TcpListener, sync::Arc};

use actix_service::Service;
use actix_web::{
    dev::{MessageBody, Server, ServiceRequest, ServiceResponse},
    error::Error,
    middleware::{Compat, Condition},
    web::{method, Data},
    HttpResponse,
};
use core::{
    config::JSON_BODY_LIMIT,
    env::env_bool,
    http::{get_addr, get_tcp_fd},
    settings::RuntimeSettings,
};
use futures::Future;
use sqlx::postgres::PgPool;
use tracing::Span;
use tracing_actix_web::{root_span, DefaultRootSpanBuilder, RootSpanBuilder, TracingLogger};

use crate::{
    error::BasicError,
    service::{self, mail, s3, upload::cleaner, ServiceData},
    translate,
};

mod cors;
mod endpoints;

fn log_ise<B: MessageBody, T>(
    request: ServiceRequest,
    srv: &T,
) -> impl Future<Output = actix_web::Result<T::Response>>
where
    T: Service<ServiceRequest, Response = ServiceResponse<B>, Error = actix_web::Error>,
{
    let uri: serde_json::Value = request.uri().to_string().into();
    let method: serde_json::Value = request.method().to_string().into();

    let future = srv.call(request);
    async {
        let mut result = future.await?;
        if result.status() == 500 {
            let response: &mut actix_web::HttpResponse<_> = result.response_mut();

            if let Some(err) = response.extensions_mut().remove::<anyhow::Error>() {
                log::error!("ISE while responding to request: {:?}", err);
                sentry::add_breadcrumb(sentry::Breadcrumb {
                    ty: "http".to_owned(),
                    category: Some("request".into()),
                    data: {
                        let mut map = sentry::protocol::Map::new();
                        map.insert("url".to_owned(), uri);
                        map.insert("method".to_owned(), method);
                        map
                    },
                    ..Default::default()
                });

                sentry::integrations::anyhow::capture_anyhow(&err);
            }
        }

        Ok(result)
    }
}

pub struct Application {
    port: u16,
    server: Option<Server>,
}

impl Application {
    pub fn new(port: u16, server: Server) -> Self {
        Self {
            port,
            server: Some(server),
        }
    }

    pub fn port(&self) -> u16 {
        self.port
    }

    // A more expressive name that makes it clear that
    // this function only returns when the application is stopped.
    pub async fn run_until_stopped(mut self) -> Result<(), std::io::Error> {
        if let Some(server) = self.server.take() {
            server.await?;
        }

        Ok(())
    }

    pub async fn stop(mut self, graceful: bool) {
        if let Some(server) = self.server.take() {
            server.stop(graceful).await
        }
    }
}

impl Drop for Application {
    fn drop(&mut self) {
        if let Some(server) = self.server.take() {
            let _ = tokio::spawn(server.stop(false));
        }
    }
}

#[actix_web::main]
pub async fn build_and_run(
    pool: PgPool,
    settings: RuntimeSettings,
    s3: Option<s3::Client>,
    gcp_key_store: Option<service::GcpAccessKeyStore>,
    gcs: Option<service::storage::Client>,
    algolia: Option<crate::algolia::Client>,
    algolia_key_store: Option<crate::algolia::SearchKeyStore>,
    jwk_verifier: Arc<crate::jwk::JwkVerifier>,
    mail_client: Option<mail::Client>,
    algolia_manager: Option<crate::algolia::Manager>,
    media_upload_cleaner: Option<cleaner::UploadCleaner>,
    google_translate: Option<translate::GoogleTranslate>,
) -> anyhow::Result<()> {
    let app = build(
        pool,
        settings,
        s3,
        gcp_key_store,
        gcs,
        algolia,
        algolia_key_store,
        jwk_verifier,
        mail_client,
        algolia_manager,
        media_upload_cleaner,
        google_translate,
    )?;
    app.run_until_stopped().await?;

    Ok(())
}

pub fn build(
    pool: PgPool,
    settings: RuntimeSettings,
    s3: Option<s3::Client>,
    gcp_key_store: Option<service::GcpAccessKeyStore>,
    gcs: Option<service::storage::Client>,
    algolia: Option<crate::algolia::Client>,
    algolia_key_store: Option<crate::algolia::SearchKeyStore>,
    jwk_verifier: Arc<crate::jwk::JwkVerifier>,
    mail_client: Option<mail::Client>,
    algolia_manager: Option<crate::algolia::Manager>,
    media_upload_cleaner: Option<cleaner::UploadCleaner>,
    google_translate: Option<translate::GoogleTranslate>,
) -> anyhow::Result<Application> {
    let local_insecure = settings.is_local();
    let api_port = settings.api_port;

    let s3 = s3.map(ServiceData::new);
    let gcp_key_store = gcp_key_store.map(ServiceData::new);
    let gcs = gcs.map(ServiceData::new);
    let algolia = algolia.map(ServiceData::new);
    let algolia_key_store = algolia_key_store.map(ServiceData::new);
    let mail_client = mail_client.map(ServiceData::new);
    let algolia_manager = algolia_manager.map(ServiceData::new);
    let media_upload_cleaner = media_upload_cleaner.map(ServiceData::new);
    let google_translate = google_translate.map(ServiceData::new);

    let server = actix_web::HttpServer::new(move || {
        let app = actix_web::App::new()
            .app_data(Data::new(pool.clone()))
            .app_data(Data::new(settings.clone()));

        let app = match s3.clone() {
            Some(s3) => app.app_data(s3),
            None => app,
        };

        let app = match gcp_key_store.clone() {
            Some(gcp_key_store) => app.app_data(gcp_key_store),
            None => app,
        };

        let app = match gcs.clone() {
            Some(gcs) => app.app_data(gcs),
            None => app,
        };

        let app = match algolia.clone() {
            Some(algolia) => app.app_data(algolia),
            None => app,
        };

        let app = match algolia_key_store.clone() {
            Some(algolia_key_store) => app.app_data(algolia_key_store),
            None => app,
        };

        let app = match mail_client.clone() {
            Some(mail_client) => app.app_data(mail_client),
            None => app,
        };

        let app = match algolia_manager.clone() {
            Some(algolia_manager) => app.app_data(algolia_manager),
            None => app,
        };

        let app = match media_upload_cleaner.clone() {
            Some(media_upload_cleaner) => app.app_data(media_upload_cleaner),
            None => app,
        };
        let app = match google_translate.clone() {
            Some(google_translate) => app.app_data(google_translate),
            None => app,
        };

        let enable_tracing_logs = env_bool("ENABLE_TRACING_LOGS");

        app.app_data(Data::from(jwk_verifier.clone()))
            .wrap(cors::get(local_insecure))
            .wrap(Condition::new(
                !enable_tracing_logs,
                Compat::new(actix_web::middleware::Logger::default()),
            ))
            .wrap(Condition::new(
                enable_tracing_logs,
                Compat::new(TracingLogger::<JigziSpanBuilder>::new()),
            ))
            .wrap_fn(log_ise)
            .app_data(
                actix_web::web::JsonConfig::default()
                    .limit(JSON_BODY_LIMIT as usize)
                    .error_handler(|_, _| bad_request_handler()),
            )
            .app_data(
                actix_web::web::QueryConfig::default().error_handler(|_, _| bad_request_handler()),
            )
            .app_data(
                actix_web::web::PathConfig::default().error_handler(|_, _| bad_request_handler()),
            )
            .external_resource(
                "google_cloud_oauth",
                "https://accounts.google.com/o/oauth2/v2/auth",
            )
            .default_service(actix_web::web::to(default_route))
            .configure(endpoints::user::configure)
            .configure(endpoints::category::configure)
            .configure(endpoints::image::configure)
            .configure(endpoints::audio::configure)
            .configure(endpoints::meta::configure)
            .configure(endpoints::jig::configure)
            .configure(endpoints::module::configure)
            .configure(endpoints::additional_resource::configure)
            .configure(endpoints::jig::report::configure)
            .configure(endpoints::jig::curation::configure)
            .configure(endpoints::course::configure)
            .configure(endpoints::admin::configure)
            .configure(endpoints::animation::configure)
            .configure(endpoints::search::configure)
            .configure(endpoints::media::configure)
            .configure(endpoints::session::configure)
            .configure(endpoints::locale::configure)
            .configure(endpoints::scheduler::configure)
            .configure(endpoints::pdf::configure)
            .configure(endpoints::badge::configure)
            .route("/", method(http::Method::GET).to(no_content_response))
    });

    // if listenfd doesn't take a TcpListener (i.e. we're not running via
    // the command above), we fall back to explicitly binding to a given
    // host:port.
    let listener = if let Some(l) = get_tcp_fd() {
        l
    } else {
        TcpListener::bind(get_addr(Some(api_port)))?
    };

    let port = listener.local_addr().unwrap().port();

    let server = server.listen(listener)?;

    Ok(Application::new(port, server.run()))
}

fn default_route() -> HttpResponse {
    HttpResponse::NotFound().json(BasicError::with_message(
        http::StatusCode::NOT_FOUND,
        "Route not found".to_owned(),
    ))
}

async fn no_content_response() -> HttpResponse {
    HttpResponse::NoContent().finish()
}

pub fn bad_request_handler() -> actix_web::Error {
    BasicError::new(http::StatusCode::BAD_REQUEST).into()
}

pub struct JigziSpanBuilder;

impl RootSpanBuilder for JigziSpanBuilder {
    fn on_request_start(request: &ServiceRequest) -> Span {
        let route = request.match_pattern().unwrap_or("default".to_string());
        let transaction_name = format!("{} {}", request.method(), route);
        let transaction_name = transaction_name.as_str();
        root_span!(request, message = transaction_name)
    }

    fn on_request_end<B>(span: Span, outcome: &Result<ServiceResponse<B>, Error>) {
        DefaultRootSpanBuilder::on_request_end(span, outcome)
    }
}
