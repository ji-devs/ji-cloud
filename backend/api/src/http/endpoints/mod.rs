pub mod additional_resource;
pub mod admin;
pub mod animation;
pub mod audio;
pub mod category;
pub mod image;
pub mod jig;
pub mod locale;
pub mod media;
pub mod meta;
pub mod module;
pub mod search;
pub mod session;
pub mod user;
