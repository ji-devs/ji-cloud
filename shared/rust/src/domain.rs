//! Types that travel over the wire.

macro_rules! into_uuid {
    ( $( $t:ty ),+ $(,)? ) => {
        $(
            impl From<$t> for uuid::Uuid {
                fn from(t: $t) -> Self {
                    t.0
                }
            }
        )+
    };
}

pub mod admin;
pub mod animation;
pub mod audio;
pub mod category;
pub mod image;
pub mod jig;
pub mod locale;
pub mod media;
pub mod meta;
pub mod search;
mod ser;
pub mod session;
pub mod user;

#[deprecated]
/// auth types (deprecated)
pub mod auth {

    #[deprecated]
    pub use super::session::AUTH_COOKIE_NAME;

    #[deprecated]
    pub use super::session::CSRF_HEADER_NAME;

    #[deprecated]
    pub use super::user::PutProfileRequest as RegisterRequest;
}

use chrono::Utc;
#[cfg(feature = "backend")]
use paperclip::actix::Apiv2Schema;
use ser::{csv_encode_uuids, deserialize_optional_field, from_csv};
use uuid::Uuid;

/// Serialize/Deserialize wrapper for Base64 encoded content.
#[cfg_attr(feature = "backend", derive(Apiv2Schema))]
#[cfg_attr(feature = "backend", openapi(empty))]
#[derive(Debug)]
pub struct Base64<T>(pub T);

impl<T: std::fmt::Display> serde::Serialize for Base64<T> {
    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: serde::Serializer,
    {
        serializer.serialize_str(&base64::encode(&self.0.to_string()))
    }
}

impl<'de, E: std::fmt::Debug, T: std::str::FromStr<Err = E>> serde::Deserialize<'de> for Base64<T> {
    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
    where
        D: serde::Deserializer<'de>,
    {
        Ok(Self(deserializer.deserialize_str(ser::FromStrVisiter(
            std::marker::PhantomData,
        ))?))
    }
}
/// Response for successfuly creating a Resource.
#[derive(Debug, serde::Serialize, serde::Deserialize, Clone)]
#[cfg_attr(feature = "backend", derive(Apiv2Schema))]
pub struct CreateResponse<T: Into<Uuid>> {
    /// The newly created resource's ID.
    pub id: T,
}

/// Represents when to publish an image.
#[derive(Copy, Clone, Eq, PartialEq, serde::Serialize, serde::Deserialize, Debug)]
#[cfg_attr(feature = "backend", derive(Apiv2Schema))]
pub enum Publish {
    /// Publish the image *at* the given time.
    At(chrono::DateTime<Utc>),
    /// Publish the image *in* the given amount of time from now.
    In(std::time::Duration),
}

impl Publish {
    /// creates an instance of `Self` that will publish "right now"
    #[must_use]
    #[allow(clippy::missing_const_for_fn)]
    pub fn now() -> Self {
        // Duration::new is const unstable
        Self::In(std::time::Duration::new(0, 0))
    }
}

impl From<Publish> for chrono::DateTime<Utc> {
    fn from(publish: Publish) -> Self {
        match publish {
            Publish::At(t) => t,
            Publish::In(d) => {
                // todo: error instead of panicing
                Utc::now() + chrono::Duration::from_std(d).expect("Really really big duration?")
            }
        }
    }
}
