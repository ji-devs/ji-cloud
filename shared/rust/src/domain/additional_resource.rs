//! Types for additional resources for JIG or Courses.

use crate::{
    api::endpoints::PathPart,
    domain::{asset::AssetId, audio::AudioId, image::ImageId, meta::ResourceTypeId, pdf::PdfId},
};
use macros::make_path_parts;
use serde::{Deserialize, Serialize};
use uuid::Uuid;

/// Wrapper type around [`Uuid`](Uuid), represents the ID of an additional resource.
#[derive(Copy, Clone, Eq, PartialEq, Serialize, Deserialize, Debug, PathPart)]
#[cfg_attr(feature = "backend", derive(sqlx::Type))]
#[cfg_attr(feature = "backend", sqlx(transparent))]
pub struct AdditionalResourceId(pub Uuid);

#[derive(Serialize, Deserialize, Debug, Clone)]
#[serde(rename_all = "camelCase")]
/// Over-the-wire representation of a JIG or Course additional resource.
pub struct AdditionalResource {
    /// The additional resources's ID.
    pub id: AdditionalResourceId,

    /// Name for additional resource
    pub display_name: String,

    /// Type of additional resource
    pub resource_type_id: ResourceTypeId,

    /// Content of additional resource
    #[serde(flatten)]
    pub resource_content: ResourceContent,
}

make_path_parts!(CreateAssetResourcePath => "/v1/additional-resource/draft");

/// Request to create a new `AdditionalResource`.
///
/// [`additional_resource::Create`](crate::api::endpoints::additional_resource::Create)
#[derive(Serialize, Deserialize, Debug)]
#[serde(rename_all = "camelCase")]
pub struct AdditionalResourceCreateRequest {
    /// Asset Id (JIG or Course) for additional resource
    #[serde(flatten)]
    pub asset_id: AssetId,

    /// Display name for additional resource
    pub display_name: String,

    /// Type of additional resource
    pub resource_type_id: ResourceTypeId,

    /// Value of additional resource
    #[serde(flatten)]
    pub resource_content: ResourceContent,
}

make_path_parts!(UpdateAssetResourcePath => "/v1/additional-resource/{}" => AdditionalResourceId);

/// Request to update an `AdditionalResource`.
///
/// [`additional_resource::Update`](crate::api::endpoints::additional_resource::Update)
#[derive(Serialize, Deserialize, Debug, Default)]
#[serde(rename_all = "camelCase")]
pub struct AdditionalResourceUpdateRequest {
    /// Asset Id (JIG or Course) for additional resource
    #[serde(flatten)]
    #[serde(default)]
    #[serde(skip_serializing_if = "Option::is_none")]
    pub asset_id: Option<AssetId>,

    /// resource display name
    #[serde(skip_serializing_if = "Option::is_none")]
    #[serde(default)]
    pub display_name: Option<String>,

    /// Type of additional  resource
    #[serde(skip_serializing_if = "Option::is_none")]
    #[serde(default)]
    pub resource_type_id: Option<ResourceTypeId>,

    /// Kind of additional resource
    #[serde(skip_serializing_if = "Option::is_none")]
    #[serde(default)]
    #[serde(flatten)]
    pub resource_content: Option<ResourceContent>,
}

make_path_parts!(GetAssetResourceDraftPath => "/v1/additional-resource/{}/draft" => AdditionalResourceId);

make_path_parts!(GetAssetResourceLivePath => "/v1/additional-resource/{}/live" => AdditionalResourceId);

make_path_parts!(DeleteAssetResourcePath => "/v1/additional-resource/{}/draft" => AdditionalResourceId);

/// Delete an `AdditionalResource` by Asset Id.
///
/// [`additional_resource::GetLive`](crate::api::endpoints::additional_resource::GetLive)
#[derive(Serialize, Deserialize, Debug, Default)]
#[serde(rename_all = "camelCase")]
pub struct AssetIdResource {
    /// Asset Id (JIG or Course) for additional resource
    #[serde(flatten)]
    #[serde(default)]
    #[serde(skip_serializing_if = "Option::is_none")]
    pub asset_id: Option<AssetId>,
}

/// Value of additional resource
#[derive(Deserialize, Serialize, Debug, Clone)]
#[serde(rename_all = "camelCase")]
pub enum ResourceContent {
    /// Additional resource kind: image
    ImageId(ImageId),
    /// Additional resource kind: audioFile
    AudioId(AudioId),
    /// Additional resource kind: link
    Link(url::Url),
    /// Additional resource kind: pdf
    PdfId(PdfId),
}

into_uuid![AdditionalResourceId];
