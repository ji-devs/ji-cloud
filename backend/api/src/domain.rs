use actix_http::body::BoxBody;
use actix_web::{
    cookie::{Cookie, SameSite},
    HttpResponse, Responder,
};

use crate::{
    db::{self, account::AccountMember},
    error,
};
use chrono::{DateTime, Utc};
use http::StatusCode;
use shared::domain::billing::AccountId;
use shared::domain::user::{UserId, UserScope};
use shared::domain::{
    category::{Category, CategoryId},
    session::AUTH_COOKIE_NAME,
};
use sqlx::PgPool;
use std::{cell::RefCell, convert::TryFrom, rc::Rc};
use std::{collections::HashMap, fmt};
use tracing::instrument;
use uuid::Uuid;

#[derive(Debug)]
pub struct RawCategory {
    pub id: Uuid,
    pub parent_id: Option<Uuid>,
    pub name: String,
    pub index: i16,
    pub created_at: DateTime<Utc>,
    pub updated_at: Option<DateTime<Utc>>,
    pub user_scopes: Vec<i16>,
}

#[derive(Debug, Clone)]
struct CategoryNode {
    pub id: Uuid,
    pub name: String,
    pub created_at: DateTime<Utc>,
    pub updated_at: Option<DateTime<Utc>>,
    pub user_scopes: Vec<i16>,
    pub children: Vec<Rc<RefCell<CategoryNode>>>,
}

impl From<CategoryNode> for Category {
    fn from(category_node: CategoryNode) -> Self {
        Self {
            id: CategoryId(category_node.id),
            name: category_node.name,
            created_at: category_node.created_at,
            updated_at: category_node.updated_at,
            user_scopes: {
                category_node
                    .user_scopes
                    .into_iter()
                    .map(|x| UserScope::try_from(x).expect("detected an invalid user scope"))
                    .collect()
            },
            children: {
                category_node
                    .children
                    .iter()
                    .map(|category| {
                        let category: CategoryNode = (**category).borrow().clone();
                        category.into()
                    })
                    .collect()
            },
        }
    }
}

#[instrument(skip_all)]
pub fn build_tree(categories: Vec<RawCategory>) -> Vec<Category> {
    let mut nodes: Vec<Rc<RefCell<CategoryNode>>> = Vec::new();
    let mut lookup: HashMap<Uuid, Rc<RefCell<CategoryNode>>> = HashMap::new();

    // Now we know this category exists in the lookup table
    for raw in categories.iter() {
        lookup.insert(
            raw.id.clone(),
            Rc::new(RefCell::new(CategoryNode {
                id: raw.id.clone(),
                name: raw.name.clone(),
                created_at: raw.created_at,
                updated_at: raw.updated_at,
                user_scopes: raw.user_scopes.clone(),
                children: Vec::new(),
            })),
        );
    }

    for raw in categories.iter() {
        let current_node = lookup
            .get(&raw.id)
            .expect("Category wasn't added to lookup map");
        match raw.parent_id {
            None => {
                // This is a root category, add it to the nodes list
                nodes.push(current_node.clone());
            }
            Some(parent_id) => {
                let parent_node = lookup
                    .get(&parent_id)
                    .expect("Parent caregory wasn't added to lookup map");
                let mut parent_node = (**parent_node).borrow_mut();
                parent_node.children.push(current_node.clone());
            }
        }
    }

    nodes
        .iter()
        .map(|node| {
            let category: CategoryNode = (**node).borrow().clone();
            let category: Category = category.into();
            category
        })
        .collect()
}

#[derive(Debug, Copy, Clone, sqlx::Type)]
#[repr(u16)]
pub enum RegistrationStatus {
    /// The user was just registered ([`POST /v1/user`](shared::api::endpoints::user::Register))
    ///
    /// This state is *skipped* via ouath (they're automatically [`Validated`](Self::Validated))
    /// In this state the user is allowed to:
    /// * Request a new verification email
    /// * Request a password reset(?)
    New = 0,

    /// The user has gone through OAuth and created a new user, or has finished email verification.
    ///
    /// In this state the user is allowed to:
    /// * Request a password reset
    /// * Complete registration via creating their profile
    /// * Delete their account
    Validated = 1,

    /// The user has *completely* finished signup, and their profile has been completed.
    ///
    /// In this state the user is allowed to do anything that their scopes allow.
    /// All users can:
    /// * Change their email (not implemented nor currently on the roadmap)
    /// * Change their password if logged in, reset their password if not.
    /// * Access the API
    /// * Change their profile details (not currently implemented)
    /// * Delete their account
    Complete = 2,
}

#[derive(Debug)]
pub struct NoContentClearAuth;

impl fmt::Display for NoContentClearAuth {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        f.write_str("No Content")
    }
}

impl Responder for NoContentClearAuth {
    type Body = BoxBody;

    fn respond_to(self, _: &actix_web::HttpRequest) -> HttpResponse<BoxBody> {
        let mut cookie = Cookie::named(AUTH_COOKIE_NAME);
        cookie.set_max_age(time::Duration::seconds(0));
        cookie.set_http_only(true);
        cookie.set_same_site(SameSite::Strict);

        actix_web::HttpResponse::build(StatusCode::NO_CONTENT)
            .content_type("application/octet-stream")
            .cookie(cookie)
            .finish()
    }
}

#[derive(Debug)]
pub enum UserAuthorization {
    SystemAdministrator,
    AccountAdministrator,
    AccountMember,
}

impl UserAuthorization {
    pub fn is_system_administrator(&self) -> bool {
        matches!(self, UserAuthorization::SystemAdministrator)
    }

    pub fn test_authorized(&self, require_account_admin: bool) -> Result<(), error::Account> {
        if self.is_authorized(require_account_admin) {
            Ok(())
        } else {
            Err(error::Account::Forbidden)
        }
    }

    pub fn is_authorized(&self, require_account_admin: bool) -> bool {
        match self {
            UserAuthorization::AccountMember if require_account_admin => false,
            _ => true,
        }
    }
}

pub async fn user_authorization(
    db: &PgPool,
    user_id: &UserId,
    account_id: &AccountId,
) -> Result<UserAuthorization, error::Account> {
    Ok(
        if db::user::has_scopes(db, *user_id, &[UserScope::Admin]).await? {
            UserAuthorization::SystemAdministrator
        } else {
            match db::account::user_account_membership(db, user_id, account_id).await? {
                Some(AccountMember::Admin) => UserAuthorization::AccountAdministrator,
                Some(AccountMember::User) => UserAuthorization::AccountMember,
                None => return Err(error::Account::Forbidden),
            }
        },
    )
}
