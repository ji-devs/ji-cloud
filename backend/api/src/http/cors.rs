use actix_web::http::{header, Method};
use core::config::CORS_ORIGINS;

pub fn get(local_insecure: bool) -> actix_cors::Cors {
    let mut cors = actix_cors::Cors::default()
        .supports_credentials()
        .allowed_methods(&[
            Method::GET,
            Method::POST,
            Method::PATCH,
            Method::DELETE,
            Method::PUT,
            Method::OPTIONS,
        ])
        .allowed_headers(&[
            header::AUTHORIZATION,
            header::CONTENT_TYPE,
            header::HeaderName::from_static("x-csrf"),
        ]);

    if local_insecure {
        cors = cors.allow_any_origin();
    } else {
        for origin in CORS_ORIGINS {
            cors = cors.allowed_origin(origin);
        }
    }

    cors
}
