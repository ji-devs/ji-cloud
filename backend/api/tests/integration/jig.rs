use http::StatusCode;
use serde_json::json;
use shared::domain::{jig::JigId, CreateResponse};

use crate::{
    fixture::Fixture,
    helpers::{initialize_server, LoginExt},
};

mod additional_resource;
mod cover;
mod draft;
mod module;

#[actix_rt::test]
async fn create_default() -> anyhow::Result<()> {
    let app = initialize_server(&[Fixture::User], &[]).await;

    let port = app.port();

    let client = reqwest::Client::new();

    let resp = client
        .post(&format!("http://0.0.0.0:{}/v1/jig", port))
        .login()
        .send()
        .await?
        .error_for_status()?;

    assert_eq!(resp.status(), StatusCode::CREATED);

    let body: CreateResponse<JigId> = resp.json().await?;

    insta::assert_json_snapshot!(body, {".id" => "[id]"});

    let resp = client
        .get(&format!("http://0.0.0.0:{}/v1/jig/{}", port, body.id.0))
        .login()
        .send()
        .await?
        .error_for_status()?;

    app.stop(false).await;

    let body: serde_json::Value = resp.json().await?;

    insta::assert_json_snapshot!(body, {".**.id" => "[id]", ".*.last_edited" => "[time_stamp]"});

    Ok(())
}

// #[actix_rt::test]
// async fn delete() -> anyhow::Result<()> {
//     let app = initialize_server(&[Fixture::User, Fixture::Jig]).await;

//     let port = app.port();

//     let client = reqwest::Client::new();

//     let resp = client
//         .delete(&format!(
//             "http://0.0.0.0:{}/v1/jig/0cc084bc-7c83-11eb-9f77-e3218dffb008",
//             port
//         ))
//         .login()
//         .send()
//         .await?
//         .error_for_status()?;

//     assert_eq!(resp.status(), StatusCode::NO_CONTENT);

//     Ok(())
// }

#[actix_rt::test]
async fn create_with_params() -> anyhow::Result<()> {
    let app = initialize_server(&[Fixture::User, Fixture::Jig], &[]).await;

    let port = app.port();

    let client = reqwest::Client::new();

    let resp = client
        .post(&format!("http://0.0.0.0:{}/v1/jig", port))
        .json(&json!({
            "modules": ["0cbfdd82-7c83-11eb-9f77-d7d86264c3bc"],
            "display_name": "test jig",
            "description": "test description",
        }))
        .login()
        .send()
        .await?
        .error_for_status()?;

    assert_eq!(resp.status(), StatusCode::CREATED);

    let body: CreateResponse<JigId> = resp.json().await?;

    app.stop(false).await;

    insta::assert_json_snapshot!(body, {".id" => "[id]", ".last_edited" => "[last_edited]"});

    Ok(())
}

#[actix_rt::test]
async fn clone() -> anyhow::Result<()> {
    let app = initialize_server(&[Fixture::User, Fixture::Jig], &[]).await;

    let port = app.port();

    let client = reqwest::Client::new();

    let resp = client
        .post(&format!(
            "http://0.0.0.0:{}/v1/jig/0cc084bc-7c83-11eb-9f77-e3218dffb008/clone",
            port
        ))
        .login()
        .send()
        .await?
        .error_for_status()?;

    assert_eq!(resp.status(), StatusCode::CREATED);

    let CreateResponse { id: JigId(id) } = resp.json().await?;

    let resp = client
        .get(&format!("http://0.0.0.0:{}/v1/jig/{}", port, id))
        .login()
        .send()
        .await?
        .error_for_status()?;

    assert_eq!(resp.status(), StatusCode::OK);

    let body: serde_json::Value = resp.json().await?;

    app.stop(false).await;

    insta::assert_json_snapshot!(
        body, {
            ".**.id" => "[id]",
            ".**.last_edited" => "[last_edited]",
            ".**.feedback_positive" => "[audio]",
            ".**.feedback_negative" => "[audio]",
            ".**.additional_resources" => "[ids]"
        }
    );

    Ok(())
}

#[actix_rt::test]
async fn get() -> anyhow::Result<()> {
    let app = initialize_server(&[Fixture::User, Fixture::Jig], &[]).await;

    let port = app.port();

    let client = reqwest::Client::new();

    let resp = client
        .get(&format!(
            "http://0.0.0.0:{}/v1/jig/0cc084bc-7c83-11eb-9f77-e3218dffb008",
            port
        ))
        .login()
        .send()
        .await?
        .error_for_status()?;

    assert_eq!(resp.status(), StatusCode::OK);

    let body: serde_json::Value = resp.json().await?;

    app.stop(false).await;

    insta::assert_json_snapshot!(
        body, {
            ".**.last_edited" => "[last_edited]",
            ".**.feedback_positive" => "[audio]",
            ".**.feedback_negative" => "[audio]"
        }
    );

    Ok(())
}

// todo: test-exhaustiveness: create a `JigBrowse` Fixture, actually test the cases (paging, jig count, etc)
#[actix_rt::test]
async fn browse_simple() -> anyhow::Result<()> {
    let app = initialize_server(&[Fixture::User, Fixture::Jig], &[]).await;

    let port = app.port();

    let client = reqwest::Client::new();

    let resp = client
        .get(&format!("http://0.0.0.0:{}/v1/jig/browse", port))
        .login()
        .send()
        .await?
        .error_for_status()?;

    assert_eq!(resp.status(), StatusCode::OK);

    let body: serde_json::Value = resp.json().await?;

    app.stop(false).await;

    insta::assert_json_snapshot!(
        body, {
            ".**.last_edited" => "[last_edited]",
            ".**.feedback_positive" => "[audio]",
            ".**.feedback_negative" => "[audio]"
        }
    );

    Ok(())
}

// todo: test-exhaustiveness: create a `JigBrowse` Fixture, actually test the cases (paging, jig count, etc)
#[actix_rt::test]
async fn browse_own_simple() -> anyhow::Result<()> {
    let app = initialize_server(&[Fixture::UserDefaultPerms, Fixture::Jig], &[]).await;

    let port = app.port();

    let client = reqwest::Client::new();

    let resp = client
        .get(&format!(
            "http://0.0.0.0:{}/v1/jig/browse?authorId=me",
            port
        ))
        .login()
        .send()
        .await?
        .error_for_status()?;

    assert_eq!(resp.status(), StatusCode::OK);

    let body: serde_json::Value = resp.json().await?;

    app.stop(false).await;

    insta::assert_json_snapshot!(
        body, {
            ".**.last_edited" => "[last_edited]",
            ".**.feedback_positive" => "[audio]",
            ".**.feedback_negative" => "[audio]"
        }
    );

    Ok(())
}

#[actix_rt::test]
async fn count() -> anyhow::Result<()> {
    let app = initialize_server(&[Fixture::UserDefaultPerms, Fixture::Jig], &[]).await;

    let port = app.port();

    let client = reqwest::Client::new();

    let resp = client
        .get(&format!("http://0.0.0.0:{}/v1/jig/count", port))
        .login()
        .send()
        .await?
        .error_for_status()?;

    assert_eq!(resp.status(), StatusCode::OK);

    let body: serde_json::Value = resp.json().await?;

    app.stop(false).await;

    insta::assert_json_snapshot!(body);

    Ok(())
}
