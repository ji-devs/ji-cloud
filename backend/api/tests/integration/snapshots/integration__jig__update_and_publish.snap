---
source: tests/integration/jig.rs
expression: body

---
{
  "id": "3a71522a-cd77-11eb-8dc1-af3e35f7c743",
  "publishedAt": null,
  "creatorId": "1f241e1b-b537-493f-a230-075cb16315be",
  "authorId": "1f241e1b-b537-493f-a230-075cb16315be",
  "authorName": "Bobby Tables",
  "likes": 0,
  "plays": 0,
  "jigFocus": "modules",
  "jigData": {
    "draftOrLive": "Draft",
    "displayName": "draft name",
    "modules": [],
    "ageRanges": [],
    "affiliations": [],
    "goals": [],
    "language": "he",
    "categories": [],
    "additionalResources": [],
    "description": "draft test description",
    "lastEdited": "2021-03-06T00:46:26.134651Z",
    "defaultPlayerSettings": {
      "direction": "rtl",
      "displayScore": false,
      "trackAssessments": false,
      "dragAssist": false
    },
    "theme": "Jigzi",
    "audioBackground": "funForKids",
    "audioEffects": {
      "feedbackPositive": [],
      "feedbackNegative": []
    },
    "privacyLevel": "unlisted",
    "locked": false,
    "otherKeywords": "",
    "translatedKeywords": ""
  },
  "adminData": {
    "rating": null,
    "blocked": false,
    "curated": false
  }
}
